Mathlib/CategoryTheory/Limits/HasLimits.lean

```diff
@@ -0,0 +1,1355 @@
+  apply hasLimitOfIso (e.invFunIdAssoc F)
+#align category_theory.limits.has_limit_of_equivalence_comp CategoryTheory.Limits.hasLimitOfEquivalenceComp
+
+-- `hasLimitCompEquivalence` and `hasLimitOfCompEquivalence`
+-- are proved in `CategoryTheory/Adjunction/Limits.lean`.
+section LimFunctor
+
+variable [HasLimitsOfShape J C]
+
+section
+
+/-- `limit F` is functorial in `F`, when `C` has all limits of shape `J`. -/
+@[simps obj]
+def lim : (J ⥤ C) ⥤ C where
+  obj F := limit F
+  map α := limMap α
+  map_id F := by
+    apply Limits.limit.hom_ext; intro j
+    erw [limMap_π, Category.id_comp, Category.comp_id]
+  map_comp α β := by
+    apply Limits.limit.hom_ext; intro j
+    erw [assoc, IsLimit.fac, IsLimit.fac, ← assoc, IsLimit.fac, assoc]; rfl
+#align category_theory.limits.lim CategoryTheory.Limits.lim
+
+end
+
+variable {G : J ⥤ C} (α : F ⟶ G)
+
+-- We generate this manually since `simps` gives it a weird name.
+@[simp]
+theorem limMap_eq_limMap : lim.map α = limMap α :=
+  rfl
+#align category_theory.limits.lim_map_eq_lim_map CategoryTheory.Limits.limMap_eq_limMap
+
+theorem limit.map_pre [HasLimitsOfShape K C] (E : K ⥤ J) :
+    lim.map α ≫ limit.pre G E = limit.pre F E ≫ lim.map (whiskerLeft E α) := by
+  ext
+  simp
+#align category_theory.limits.limit.map_pre CategoryTheory.Limits.limit.map_pre
+
+theorem limit.map_pre' [HasLimitsOfShape K C] (F : J ⥤ C) {E₁ E₂ : K ⥤ J} (α : E₁ ⟶ E₂) :
+    limit.pre F E₂ = limit.pre F E₁ ≫ lim.map (whiskerRight α F) := by
+  ext1; simp [← category.assoc]
+#align category_theory.limits.limit.map_pre' CategoryTheory.Limits.limit.map_pre'
+
+theorem limit.id_pre (F : J ⥤ C) : limit.pre F (𝟭 _) = lim.map (Functor.leftUnitor F).inv := by 
+  aesop_cat
+#align category_theory.limits.limit.id_pre CategoryTheory.Limits.limit.id_pre
+
+theorem limit.map_post {D : Type u'} [Category.{v'} D] [HasLimitsOfShape J D] (H : C ⥤ D) :
+    /- H (limit F) ⟶ H (limit G) ⟶ limit (G ⋙ H) vs
+     H (limit F) ⟶ limit (F ⋙ H) ⟶ limit (G ⋙ H) -/
+    H.map (limMap α) ≫ limit.post G H = limit.post F H ≫ limMap (whiskerRight α H) := by
+  ext
+  simp only [whiskerRight_app, limMap_π, assoc, limit.post_π_assoc, limit.post_π, ← H.map_comp]
+#align category_theory.limits.limit.map_post CategoryTheory.Limits.limit.map_post
+
+/-- The isomorphism between
+morphisms from `W` to the cone point of the limit cone for `F`
+and cones over `F` with cone point `W`
+is natural in `F`.
+-/
+def limYoneda :
+    lim ⋙ yoneda ⋙ (whiskeringRight _ _ _).obj uliftFunctor.{u₁} ≅ CategoryTheory.cones J C := by
+  refine 
+    NatIso.ofComponents (fun F => NatIso.ofComponents (fun W => limit.homIso F (unop W))  ?_) ?_ 
+  · intro X Y f 
+    dsimp [yoneda,uliftFunctor,cones]
+    funext g 
+    dsimp [const,cones]
+    apply NatTrans.ext
+    aesop_cat
+  · intro X Y f
+    dsimp [yoneda,uliftFunctor]
+    ext Z
+    dsimp 
+    funext g 
+    dsimp [const,cones]
+    apply NatTrans.ext
+    aesop_cat
```

Comment:
```suggestion
def limYoneda :
    lim ⋙ yoneda ⋙ (whiskeringRight _ _ _).obj uliftFunctor.{u₁} ≅ CategoryTheory.cones J C :=
  NatIso.ofComponents (fun F => NatIso.ofComponents (fun W => limit.homIso F (unop W))
    <| by intros ; funext ; aesop_cat) <| by intros ; ext ; funext ; aesop_cat
```

If we add `funext` to the `aesop_cat` rule set, then I think `aesop_cat` would be able to handle both naturality goals.

Reply:
Both are golfed